core/src/core/transaction.rs

```diff
@@ -831,8 +844,14 @@ impl Transaction {
 		Transaction {
 			offset: BlindingFactor::zero(),
 			body: Default::default(),
+			hash_cache: None,
 		}
 	}
+	fn recompute_hash(mut self) -> Transaction {
```

Comment:
This is going to be biggest concern here I think - making sure we have covered all cases of mutating existing data and _guaranteeing_ we are not using a stale hash somewhere.

Reply:
Yep!

The best way to do this is going to be to make the types non pub (only gettable via getter which provides immutable ref), or where mut is needed, to make it take a callback, e.g.

```rust
transaction.outputs_mut(|outputs| {
   outputs.push(x)
});

impl Transaction {
  fn outputs_mut<F, R>(&mut self, f: F)-> where F: Fn(&mut Vec<Output>) -> R {
    let r = f(&mut self.outputs);
    self.recompute_hashes();
    r
  } 
}
```